File: ruby/dialog/listdialog.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

require 'dialoghelper'

class ListDialog < NSObject
  include DialogHelper
  attr_accessor :delegate, :prefix
  ib_outlet :window, :table, :updateButton, :closeButton, :searchText
  
  def initialize
    @prefix = 'listDialog'
    @list = []
    @flist = nil
    @filter = nil
    @sort_key = 1
    @sort_order = :descent
  end
  
  def start
    NSBundle.loadNibNamed_owner('ListDialog', self)
    load_window_state
    @table.setDoubleAction(:onJoin)
    @window.makeFirstResponder(@updateButton)
    show
  end
  
  def show
    @window.setTitle("Channel List - #{@delegate.name}")
    @window.makeKeyAndOrderFront(self)
  end
  
  def close
    @delegate = nil
    @window.close
  end
  
  def windowWillClose(sender)
    save_window_state
    fire_event('onClose')
  end
  
  ib_action :onClose
  def onClose(sender)
    @window.close
  end
  
  ib_action :onUpdate
  def onUpdate(sender)
    fire_event('onUpdate')
  end
  
  ib_action :onJoin
  def onJoin(sender)
    i = @table.selectedRows[0]
    if i
      if @filter.nil?
        list = @list
      else
        build_filtered_list
        list = @flist
      end
      item = list[i]
      if item
        fire_event('onJoin', item[0])
      end
    end
  end
  
  ib_action :onSearchTextChanged
  def onSearchTextChanged(sender)
    filter = sender.stringValue.to_s
    @filter = filter.empty? ? nil : Regexp.compile(Regexp.escape(filter), Regexp::IGNORECASE)
    @flist = nil
    reload_table
  end
  
  def tableView_didClickTableColumn(table, col)
    i = case col.identifier
      when 'chname'; 0
      when 'count'; 1
      else; 2
    end
    if @sort_key == i
      @sort_order = @sort_order == :ascent ? :descent : :ascent
    else
      @sort_key = i
      @sort_order = :ascent
    end
    @flist = nil
    sort
    reload_table
  end
  
  def clear
    @list = []
    @flist = nil
    reload_table
  end
  
  def sort
    @list.sort! do |a,b|
      row_compare a, b
    end
  end

  def sorted_insert(ary, item)
    # do a binary search
    # once the range hits a length of 5 (arbitrary)
    # switch to linear search
    head = 0
    tail = ary.size
    while tail - head > 5
      pivot = (head + tail) / 2
      if row_compare(ary[pivot], item) > 0
        tail = pivot
      else
        head = pivot
      end
    end
    head.upto(tail-1) do |idx|
      if row_compare(ary[idx], item) > 0
        ary.insert idx, item
        return
      end
    end
    ary.insert tail, item
  end

  def row_compare(a, b)
    if @sort_key == 1
      res = a[1] <=> b[1]
    else
      res = a[@sort_key].casecmp b[@sort_key]
    end
    if res == 0
      if @sort_key == 0
        res = a[1] <=> b[1]
      else
        res = a[0].casecmp b[0]
      end
    end
    if @sort_order == :ascent
      res
    else
      - res
    end
  end
  
  def add_item(item)
    sorted_insert @list, item
    sorted_insert @flist, item if item_matches_filter?(item) unless @flist.nil?
    note_number_of_rows_changed
  end
  
  def reload_table
    @table.reloadData
  end

  def note_number_of_rows_changed
    @table.noteNumberOfRowsChanged
  end
  
  def build_filtered_list
    return if @flist
    @flist = @list.select {|i| item_matches_filter?(i) }
  end

  def item_matches_filter?(row)
    row[0] =~ @filter || row[2] =~ @filter
  end
  
  def numberOfRowsInTableView(sender)
    if @filter.nil?
      @list.size
    else
      build_filtered_list
      @flist.size
    end
  end
  
  def tableView_objectValueForTableColumn_row(sender, col, row)
    if @filter.nil?
      list = @list
    else
      build_filtered_list
      list = @flist
    end
    
    m = list[row]
    case col.identifier
      when 'chname'; m[0]
      when 'count'; m[1].to_s
      else; m[2]
    end
  end
  
  private

  def load_window_state
    if c = preferences.load_window('channel_list_window')
      frame = NSRect.from_dic(c[:window])
      set_table_header_settings(@table, c[:tablecols])
    else
      frame = NSRect.from_center(NSScreen.screens[0].visibleFrame.center, 500, 400)
    end
    @window.setFrame_display(frame, false)
  end

  def save_window_state
    if @window
      c = {
        :window => @window.frame.to_dic,
        :tablecols => get_table_header_settings(@table),
      }
      preferences.save_window('channel_list_window', c)
    end
  end
end

File: Classes/Dialogs/ListDialog.m
// LimeChat is copyrighted free software by Satoshi Nakagawa <psychs AT limechat DOT net>.
// You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL.txt).

#import "ListDialog.h"
#import "Preferences.h"
#import "NSDictionaryHelper.h"


@implementation ListDialog
{
    NSMutableArray* _list;
    NSMutableArray* _filteredList;
}

- (id)init
{
    self = [super init];
    if (self) {
        [NSBundle loadNibNamed:@"ListDialog" owner:self];

        _list = [NSMutableArray new];
        _sortKey = 1;
        _sortOrder = NSOrderedDescending;
    }
    return self;
}

- (void)start
{
    [_table setDoubleAction:@selector(onJoin:)];

    [self show];
}

- (void)show
{
    if (![self.window isVisible]) {
        NSDictionary* dic = [Preferences loadWindowStateWithName:@"channel_list_window"];
        if (dic) {
            NSDictionary* win = [dic objectForKey:@"window"];
            NSArray* cols = [dic objectForKey:@"tablecols"];

            double x = [win doubleForKey:@"x"];
            double y = [win doubleForKey:@"y"];
            double w = [win doubleForKey:@"w"];
            double h = [win doubleForKey:@"h"];
            [self.window setFrame:NSMakeRect(x, y, w, h) display:NO];

            int i = 0;
            for (NSNumber* n in cols) {
                [[[_table tableColumns] objectAtIndex:i] setWidth:[n doubleValue]];
                ++i;
            }
        }
        else {
            [self.window center];
        }
    }

    [self.window makeKeyAndOrderFront:nil];
}

- (void)close
{
    [self.window close];
}

- (void)clear
{
    [_list removeAllObjects];
    _filteredList = nil;

    [self reloadTable];
}

- (void)addChannel:(NSString*)channel count:(int)count topic:(NSString*)topic
{
    NSArray* item = [NSArray arrayWithObjects:channel, @(count), topic, nil];

    NSString* filter = [_filterText stringValue];
    if (filter.length) {
        if (!_filteredList) {
            _filteredList = [NSMutableArray new];
        }

        if ([channel rangeOfString:filter options:NSCaseInsensitiveSearch].location != NSNotFound
            || [topic rangeOfString:filter options:NSCaseInsensitiveSearch].location != NSNotFound) {
            [self sortedInsert:item inArray:_filteredList];
        }
    }

    [self sortedInsert:item inArray:_list];
    [self reloadTable];
}

- (void)reloadTable
{
    [_table reloadData];
}

static NSInteger compareItems(NSArray* self, NSArray* other, void* context)
{
    ListDialog* dialog = (__bridge ListDialog*)context;
    int key = dialog.sortKey;
    NSComparisonResult order = dialog.sortOrder;

    NSString* mine = [self objectAtIndex:key];
    NSString* others = [other objectAtIndex:key];

    NSComparisonResult result;
    if (key == 1) {
        result = [mine compare:others];
    }
    else {
        result = [mine caseInsensitiveCompare:others];
    }

    if (order == NSOrderedDescending) {
        return - result;
    }
    else {
        return result;
    }
}

- (void)sort
{
    [_list sortUsingFunction:compareItems context:(void*)self];
}

- (void)sortedInsert:(NSArray*)item inArray:(NSMutableArray*)ary
{
    const int THRESHOLD = 5;
    int left = 0;
    int right = ary.count;

    while (right - left > THRESHOLD) {
        int pivot = (left + right) / 2;
        if (compareItems([ary objectAtIndex:pivot], item, (__bridge void*)self) == NSOrderedDescending) {
            right = pivot;
        }
        else {
            left = pivot;
        }
    }

    for (int i=left; i<right; ++i) {
        if (compareItems([ary objectAtIndex:i], item, (__bridge void*)self) == NSOrderedDescending) {
            [ary insertObject:item atIndex:i];
            return;
        }
    }

    [ary insertObject:item atIndex:right];
}

#pragma mark - Actions

- (void)onClose:(id)sender
{
    [self.window close];
}

- (void)onUpdate:(id)sender
{
    if ([_delegate respondsToSelector:@selector(listDialogOnUpdate:)]) {
        [_delegate listDialogOnUpdate:self];
    }
}

- (void)onJoin:(id)sender
{
    NSArray* ary = _list;
    NSString* filter = [_filterText stringValue];
    if (filter.length) {
        ary = _filteredList;
    }

    NSIndexSet* indexes = [_table selectedRowIndexes];
    for (NSUInteger i=[indexes firstIndex]; i!=NSNotFound; i=[indexes indexGreaterThanIndex:i]) {
        NSArray* item = [ary objectAtIndex:i];
        if ([_delegate respondsToSelector:@selector(listDialogOnJoin:channel:)]) {
            [_delegate listDialogOnJoin:self channel:[item objectAtIndex:0]];
        }
    }
}

- (void)onSearchFieldChange:(id)sender
{
    _filteredList = nil;

    NSString* filter = [_filterText stringValue];
    if (filter.length) {
        NSMutableArray* ary = [NSMutableArray new];
        for (NSArray* item in _list) {
            NSString* channel = [item objectAtIndex:0];
            NSString* topic = [item objectAtIndex:2];
            if ([channel rangeOfString:filter options:NSCaseInsensitiveSearch].location != NSNotFound
                || [topic rangeOfString:filter options:NSCaseInsensitiveSearch].location != NSNotFound) {
                [ary addObject:item];
            }
        }
        _filteredList = ary;
    }

    [self reloadTable];
}

- (BOOL)loadWindowState
{
    NSDictionary* dic = [Preferences loadWindowStateWithName:@"channel_list_window"];
    if (!dic) return NO;

    NSDictionary* win = [dic objectForKey:@"window"];
    NSArray* cols = [dic objectForKey:@"tablecols"];

    double x = [win doubleForKey:@"x"];
    double y = [win doubleForKey:@"y"];
    double w = [win doubleForKey:@"w"];
    double h = [win doubleForKey:@"h"];
    [self.window setFrame:NSMakeRect(x, y, w, h) display:NO];

    int i = 0;
    for (NSNumber* n in cols) {
        [[[_table tableColumns] objectAtIndex:i] setWidth:[n doubleValue]];
        ++i;
    }

    return YES;
}

- (void)saveWindowState
{
    NSMutableDictionary* win = [NSMutableDictionary dictionary];
    NSRect rect = self.window.frame;
    [win setDouble:rect.origin.x forKey:@"x"];
    [win setDouble:rect.origin.y forKey:@"y"];
    [win setDouble:rect.size.width forKey:@"w"];
    [win setDouble:rect.size.height forKey:@"h"];

    NSMutableArray* cols = [NSMutableArray array];
    for (NSTableColumn* col in [_table tableColumns]) {
        [cols addObject:@(col.width)];
    }

    NSDictionary* dic = [NSDictionary dictionaryWithObjectsAndKeys:win, @"window", cols, @"tablecols", nil];
    [Preferences saveWindowState:dic name:@"channel_list_window"];
}

#pragma mark - NSTableView Delegate

- (NSInteger)numberOfRowsInTableView:(NSTableView *)sender
{
    if (_filteredList) {
        return _filteredList.count;
    }
    return _list.count;
}

- (id)tableView:(NSTableView *)sender objectValueForTableColumn:(NSTableColumn *)column row:(NSInteger)row
{
    NSArray* ary = _filteredList ?: _list;
    NSArray* item = [ary objectAtIndex:row];
    NSString* col = [column identifier];

    if ([col isEqualToString:@"chname"]) {
        return [item objectAtIndex:0];
    }
    else if ([col isEqualToString:@"count"]) {
        return [item objectAtIndex:1];
    }
    else {
        return [item objectAtIndex:2];
    }
}

- (void)tableView:(NSTableView *)tableView didClickTableColumn:(NSTableColumn *)column
{
    int i;
    NSString* col = [column identifier];
    if ([col isEqualToString:@"chname"]) {
        i = 0;
    }
    else if ([col isEqualToString:@"count"]) {
        i = 1;
    }
    else {
        i = 2;
    }

    if (_sortKey == i) {
        _sortOrder = - _sortOrder;
    }
    else {
        _sortKey = i;
        _sortOrder = (_sortKey == 1) ? NSOrderedDescending : NSOrderedAscending;
    }

    [self sort];

    if (_filteredList) {
        [self onSearchFieldChange:nil];
    }

    [self reloadTable];
}

#pragma mark - NSWindow Delegate

- (void)windowWillClose:(NSNotification*)note
{
    [self saveWindowState];

    if ([_delegate respondsToSelector:@selector(listDialogWillClose:)]) {
        [_delegate listDialogWillClose:self];
    }
}

@end
